File: Example/PayTPV Objc Example/CardsTableViewController.m
//
//  CardsTableViewController.m
//  PAYTPV Objc Example
//
//  Created by Mihail Cristian Dumitru on 8/21/17.
//  Copyright © 2017 PAYTPV. All rights reserved.
//

#import "PAYTPV/PAYTPV.h"

#import "CardsTableViewController.h"
#import "AddCardController.h"

#import "ModelStore.h"
#import "Models.h"
#import "AlertBuilder.h"

@interface CardsTableViewController () <AddCardControllerDelegate>

#pragma mark - Properties

@property (nonatomic, copy) NSArray *cards;

@end

@implementation CardsTableViewController

#pragma mark - Lifecycle

- (void)viewDidLoad {
    [super viewDidLoad];

    [self setup];
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];

    [self loadRemoteData];
}

#pragma mark - Setup

- (void)setup {
    self.title = @"Cards";

    self.navigationItem.rightBarButtonItems = @[[self addButtonItem],
                                                [self editButtonItem]];
}

#pragma mark - Helpers

- (UIBarButtonItem *)addButtonItem {
    return [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemAdd target:self action:@selector(onAddButtonPressed)];
}

- (void)onAddButtonPressed {
    AddCardController *controller = [AddCardController new];
    controller.delegate = self;
    [self presentViewController:controller animated:true completion:nil];
}

- (void)loadRemoteData {
    [ModelStore cards:^(NSArray * _Nonnull cards) {
        self.cards = cards;
        [self reloadData];
    }];
}

- (void)reloadData {
    [self.tableView reloadData];
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.cards.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"CardCell"];
    if (cell == nil) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleValue1 reuseIdentifier:@"CardCell"];
    }

    Card *card = [self.cards objectAtIndex:indexPath.row];

    cell.textLabel.text = card.name;

    return cell;
}

- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath {
    return YES;
}

- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath {
    if (editingStyle == UITableViewCellEditingStyleDelete) {

        // add a loading indicator
        AlertBuilder *loader = [AlertBuilder new];
        [loader setTitle:@"Removing card"];
        [loader addLoadingIndicator];
        [loader presentInController:self animated:true completion:nil];

        Card *selectedCard = [self.cards objectAtIndex:indexPath.row];

        // call the PAYTPV api to remove the user from the system
        PTPVUser *user = [[PTPVUser alloc] initWithIdUser:selectedCard.idUser tokenUser:selectedCard.tokenUser];
        [[PTPVAPIClient sharedClient] removeUser:user completion:^(NSError * _Nullable error) {
            [loader dismissAnimated:true completion:nil];

            if (error != nil) {
                AlertBuilder *alert = [AlertBuilder new];
                [alert setTitle:error.localizedDescription];
                [alert addActionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil];
                [alert presentInController:self animated:true completion:nil];
                return;
            }

            // remove the user from the backend
            [ModelStore removeCard:selectedCard completion:^{
                [self loadRemoteData];
            }];
        }];

    } else if (editingStyle == UITableViewCellEditingStyleInsert) {
        // Create a new instance of the appropriate class, insert it into the array, and add a new row to the table view
    }
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    if (self.delegate) {
        [tableView deselectRowAtIndexPath:indexPath animated:true];

        Card *selectedCard = [self.cards objectAtIndex:indexPath.row];

        [self.delegate cardsTableController:self didSelectCard:selectedCard];
    }
}

#pragma mark - AddCardControllerDelegate

- (void)addCardController:(AddCardController *)controller didFinishAddingCard:(Card *)card {
    [self reloadData];
}

@end

File: Example/PayTPV Swift Example/CardsTableViewController.swift
//
//  CardsTableViewController.swift
//  PAYTPV Swift Example
//
//  Created by Mihail Cristian Dumitru on 8/17/17.
//  Copyright © 2017 PAYTPV. All rights reserved.
//

import UIKit
import PAYTPV

@objc protocol CardsTableControllerDelegate {
    func cardsTableController(_ controller: CardsTableViewController, didSelectCard card: Card)
}

class CardsTableViewController: UITableViewController {

    // MARK: - Properties

    weak var delegate: CardsTableControllerDelegate?

    var cards: [Card] = []

    var addButtonItem: UIBarButtonItem {
        return UIBarButtonItem(barButtonSystemItem: .add, target: self, action: #selector(self.onAddButtonPressed))
    }

    // MARK: - Lifecycle

    override func viewDidLoad() {
        super.viewDidLoad()

        self.setup()
    }

    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)

        self.loadRemoteData()
    }

    // MARK: - Setup

    private func setup() {
        self.title = "Cards"

        self.navigationItem.rightBarButtonItems = [self.addButtonItem, self.editButtonItem]
    }

    // MARK: - Helpers

    func onAddButtonPressed() {
        let addCardController = AddCardController()
        addCardController.delegate = self
        self.present(addCardController, animated: true, completion: nil)
    }

    func loadRemoteData() {
        ModelStore.getCards { (cards) in
            self.cards = cards
            self.reloadData()
        }
    }

    func reloadData() {
        self.tableView.reloadData()
    }

    // MARK: - Table view data source

    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.cards.count
    }

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "CardCell") ?? UITableViewCell(style: .value1, reuseIdentifier: "CardCell")

        let card = self.cards[indexPath.row]

        cell.textLabel?.text = card.name

        return cell
    }

    override func tableView(_ tableView: UITableView, canEditRowAt indexPath: IndexPath) -> Bool {
        return true
    }

    override func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCellEditingStyle, forRowAt indexPath: IndexPath) {
        if editingStyle == .delete {

            // add a loading indicator
            let loader = AlertBuilder()
                .set(title: "Removing card")
                .addLoadingIndicator()
                .present(sender: self)

            let selectedCard = self.cards[indexPath.row]

            // call the PAYTPV api to remove the user from the system
            let user = PTPVUser(idUser: selectedCard.id, tokenUser: selectedCard.token)
            PTPVAPIClient.shared().remove(user, completion: { (error) in
                loader.dismiss()

                if let error = error {
                    let _ = AlertBuilder()
                        .set(message: error.localizedDescription)
                        .add(actionWithTitle: "OK")
                        .present(sender: self)
                    return;
                }

                // remove the user from the backend
                ModelStore.remove(card: selectedCard, completion: { (_) in
                    self.loadRemoteData()
                })
            })

        } else if editingStyle == .insert {
            // Create a new instance of the appropriate class, insert it into the array, and add a new row to the table view
        }
    }

    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)

        let selectedCard = self.cards[indexPath.row]

        self.delegate?.cardsTableController(self, didSelectCard: selectedCard)
    }

}

// MARK: - AddCardDelegate

extension CardsTableViewController: AddCardDelegate {

    func addCardController(_ controller: AddCardController, didFinishAddingCard card: Card) {
        self.reloadData()
    }

}
